test/unit/math/prim/fun/ternary_scalar_tester.hpp

```diff
@@ -0,0 +1,288 @@
+#ifndef TEST_UNIT_MATH_PRIM_FUN_TERNARY_SCALAR_TESTER_HPP
+#define TEST_UNIT_MATH_PRIM_FUN_TERNARY_SCALAR_TESTER_HPP
+#include <stan/math/prim.hpp>
+#include <gtest/gtest.h>
+
+namespace stan {
+namespace test {
+
+/**
+ * Implementation function which checks that the ternary vectorisation
+ * framework returns the same value as the function with scalar inputs,
+ * for all valid combinations of scalar/vector/nested vector.
+ *
+ * Specialization for use with matrix inputs
+ *
+ * @tparam F Type of functor to apply.
+ * @tparam T1 Type of first matrix.
+ * @tparam T2 Type of second matrix.
+ * @tparam T3 Type of third matrix.
+ * @param x First matrix input to which operation is applied.
+ * @param y Second matrix input to which operation is applied.
+ * @param z Third matrix input to which operation is applied.
+ * @param f functor to apply to inputs.
+ */
+template <typename F, typename T1, typename T2, typename T3,
+          require_all_not_vector_t<T1, T2, T3>* = nullptr>
+void ternary_scalar_tester_impl(const F& f, const T1& x, const T2& y,
+                                const T3& z) {
+  auto mat_mat_mat = math::eval(f(x, y, z));
+  auto mat_mat_scal = math::eval(f(x, y, z(0)));
+  auto mat_scal_scal = math::eval(f(x, y(0), z(0)));
+  auto mat_scal_mat = math::eval(f(x, y(0), z));
+  auto scal_scal_mat = math::eval(f(x(0), y(0), z));
+  auto scal_mat_scal = math::eval(f(x(0), y, z(0)));
+  auto scal_mat_mat = math::eval(f(x(0), y, z));
+
+  for (int i = 0; i < x.size(); ++i) {
+    EXPECT_FLOAT_EQ(f(x(i), y(i), z(i)), mat_mat_mat(i));
+    EXPECT_FLOAT_EQ(f(x(i), y(i), z(0)), mat_mat_scal(i));
+    EXPECT_FLOAT_EQ(f(x(i), y(0), z(0)), mat_scal_scal(i));
+    EXPECT_FLOAT_EQ(f(x(i), y(0), z(i)), mat_scal_mat(i));
+    EXPECT_FLOAT_EQ(f(x(0), y(0), z(i)), scal_scal_mat(i));
+    EXPECT_FLOAT_EQ(f(x(0), y(i), z(0)), scal_mat_scal(i));
+    EXPECT_FLOAT_EQ(f(x(0), y(i), z(i)), scal_mat_mat(i));
+  }
+  plain_type_t<T1> x_zero;
+  plain_type_t<T2> y_zero;
+  plain_type_t<T3> z_zero;
+  EXPECT_THROW(f(x_zero, y, z), std::invalid_argument);
+  EXPECT_THROW(f(x, y_zero, z), std::invalid_argument);
+  EXPECT_THROW(f(x, y, z_zero), std::invalid_argument);
```

Comment:
I think if we want to be able to return a zero-sized type for Eigen types, then we'd have to return the plain type rather than the expression result. Eigen throws for zero-sized inputs to `CwiseTernaryOp`, and I can't instantiate a zero-sized expression to match the return type for auto.

`apply_scalar_binary` would also need to be updated to have this behaviour, so might be better making a separate issue